### PR TITLE
Typescript decorator factories

### PR DESCRIPTION
**Pull request by Aboisier**
Fix for #1859. 

Currently, having a decorator factory in a constructor breaks the highlighting. 

Example:

```ts
constructor(@decorator() private bar: Bar) { }
```

Is rendered as (notice the `</span>` after `@decorator(` )
```html
<span class="hljs-keyword">constructor</span>(<span class="hljs-params">@decorator(</span>) private bar: Bar) { }
```

It should be rendered as 
```html
<span class="hljs-keyword">constructor</span>(<span class="hljs-params"><span class="hljs-meta">@decorator</span>() <span class="hljs-keyword">private</span> bar: Bar</span>) { }
```

There are also a couple more scenarios to take into account:
```ts
constructor(@decorator('Hello') private bar: Bar) { }
constructor(@decorator(123) private bar: Bar) { }
constructor(@decorator(true) private bar: Bar) { }
constructor(@decorator(someFunction('Hello')) private bar: Bar) { }
function(@decorator() param: Param) { }
```

This pull request adds decorators and arguments. These constructs may be contained in the root, in functions definitions, and in constructors. I added a test with different scenarios. I reduced duplication by creating the PARAMS object. Finally, the current decorators definition did not allow decorators names to start with `$` or `_`, which is a valid. I also fixed that.

Thanks for considering this fix!
Antoine

### CODE REVIEW

**Comment by marcoscaceres** [comment]
@Aboisier, while we wait, would you be so kind as to add your “improvement:” to CHANGES.md just under `## Master`. Thanks in advance!

**Comment by egor-rogov** [comment]
Looks just fine!

File: CHANGES.md [hunk 1]
```diff
@@ -42,6 +42,7 @@ Improvements:
 - *Matlab*: transpose operators and double quote strings, by [JohnC32][] and [Egor Rogov][]
 - Various documentation typos and improvemets by [Jimmy Wärting][], [Lutz Büch][], [bcleland][]
 - *Cmake* updated with new keywords and commands by [Deniz Bahadir][]
+- *TypeScript* decorator factories highlighting fixed by [Antoine Boisier-Michaud][]
```
Comment (marcoscaceres) [inline CHANGES.md]: You might need to rebase, as 9.13.0 already shipped.

**Comment by marcoscaceres** [comment]
Will get this out in a few weeks as part of 9.14.0.

**Pull request merged**